File: Python3/프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.py
def solution(n, tops):
    MOD = 10007
    if n == 1:
        return 3 if tops[0] == 1 else 2

    dp0 = 1
    dp1 = 3 if tops[0] == 1 else 2  # dp[1]

    # i = 2 .. 2n
    for i in range(2, 2 * n + 1):
        if (i & 1) == 1 and tops[i // 2] == 1:
            cur = (dp1 * 2 + dp0) % MOD
        else:
            cur = (dp1 + dp0) % MOD
        dp0, dp1 = dp1, cur

    return dp1

File: C++/프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.cpp
#include <vector>
using namespace std;

int solution(int n, vector<int> tops) {
    const int MOD = 10007;

    if (n == 1) return (tops[0] == 1) ? 3 : 2;

    int dp0 = 1;
    int dp1 = (tops[0] == 1) ? 3 : 2; // dp[1]

    for (int i = 2; i <= 2 * n; i++) {
        int cur;
        if ((i & 1) == 1 && tops[i / 2] == 1) {
            cur = (dp1 * 2 + dp0) % MOD;
        } else {
            cur = (dp1 + dp0) % MOD;
        }
        dp0 = dp1;
        dp1 = cur;
    }
    return dp1;
}
